### PR TITLE
8304074: [JMX] Add an approximation of total bytes allocated on the Java heap by the JVM

### PR DESCRIPTION
**Pull request by phohensee**
I'd like to backport the definition and implementation of com.sun.management.ThreadMXBean.getTotalThreadAllocatedBytes to 8u. The backport CSR is [JDK-8320375](https://bugs.openjdk.org/browse/JDK-8320375). A follow-up bugfix backport of [JDK-8313081](https://bugs.openjdk.org/browse/JDK-8313081) will be done following this backport. The combined backports have been in production at Amazon for two months with no issues. The backport uses the reserved1 slot in jmm.h in order to preserve binary compatibility with 8u. Per current policy, there is no update to JMM_VERSION in jmm.h and the new method is marked
```
@since 8u412
```
Aside from file relocation and context differences, relative to the 11u backport the MonitoringSupport_lock definition macro changes, and the reserved1 rather than reserved6 jmm slot is used. SMR doesn't exist in 8u, so jmm_GetTotalThreadAllocated bytes attempts to lock Threads_lock instead, and if the Threads_lock is already locked, the previous return value is returned. HotSpotThreadImpl.java doesn't exist in 8u, so the hunk associated with it is dropped.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8320375](https://bugs.openjdk.org/browse/JDK-8320375) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8304074](https://bugs.openjdk.org/browse/JDK-8304074) needs maintainer approval

### Issues
 * [JDK-8304074](https://bugs.openjdk.org/browse/JDK-8304074): [JMX] Add an approximation of total bytes allocated on the Java heap by the JVM (**Enhancement** - P4 - Approved)
 * [JDK-8320375](https://bugs.openjdk.org/browse/JDK-8320375): [JMX] Add an approximation of total bytes allocated on the Java heap by the JVM (**CSR**)

### Reviewers
 * [Volker Simonis](https://openjdk.org/census#simonis) (@simonis - **Reviewer**)
 * [Andrew John Hughes](https://openjdk.org/census#andrew) (@gnu-andrew - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk8u-dev.git pull/392/head:pull/392` \
`$ git checkout pull/392`

Update a local copy of the PR: \
`$ git checkout pull/392` \
`$ git pull https://git.openjdk.org/jdk8u-dev.git pull/392/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 392`

View PR using the GUI difftool: \
`$ git pr show -t 392`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk8u-dev/pull/392.diff">https://git.openjdk.org/jdk8u-dev/pull/392.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk8u-dev/pull/392#issuecomment-1826882257)

### CODE REVIEW

**Review by simonis — CHANGES_REQUESTED** [review]
I didn't read the PR comment carefully (my fault :) so here's my list of changes with regards to the corresponding jdk11 change set (see https://github.com/openjdk/jdk11u-dev/pull/2286). I'll provide them here in the hope to simplify the review for others:

- in jdk8 we have two `jmm.h` files, one in `jdk/src/share/javavm/export/jmm.h` and one in `hotspot/src/share/vm/services/jmm.h`. The changes to both files are the same.
- There's no `HotSpotThreadImpl.java` so the function added in the base class `ThreadImpl.java` has to be public in jdk8u.
- there is no `jdk/make/mapfiles/libmanagement/mapfile-vers` in `jdk11u` and later.
- there's no `JavaThreadIteratorWithHandle`/`ThreadsListHandle` in jdk8 that's why we have to use the `Threads_lock` while iterating over the threads. It's also the reason why parts of the comments in `jmm_GetTotalThreadAllocatedMemory()` became obsolete and have been removed.

@phohensee, is my assumption correct that you also use this already in production within Amazon and haven't seen any negative impact (particularly not due to the usage of the `Threads_lock` in `jmm_GetTotalThreadAllocatedMemory()` ?

**Comment by phohensee** [comment]
Thanks for the quick review, Volker! Re your comments:

Yes, there are two jmm.h files with identical changes. I missed pointing that out.
HotSpotThreadImpl.java was mentioned in the PR comment.
Yes, there is no jdk/make/mapfiles/libmanagement/mapfile-vers in jdk11u and later. I missed pointing that out too.
Lack of SMR was mentioned in the PR comment.

File: jdk/make/mapfiles/libmanagement/mapfile-vers [hunk 1]
```diff
@@ -89,6 +89,7 @@ SUNWprivate_1.1 {
 	    Java_sun_management_ThreadImpl_getThreadUserCpuTime1;
 	    Java_sun_management_ThreadImpl_getThreadAllocatedMemory0;
 	    Java_sun_management_ThreadImpl_getThreadAllocatedMemory1;
+            Java_sun_management_ThreadImpl_getTotalThreadAllocatedMemory;
```
[the remark this answers is not in the document]
Comment (phohensee) [reply jdk/make/mapfiles/libmanagement/mapfile-vers]: Fixed.

**Review by simonis — APPROVED** [review]
Thanks for the addition information Paul. Looks good now.

**Comment by phohensee** [comment]
Pre-submit failures appear unrelated.

**Comment by phohensee** [comment]
/approval request Backport of a purely additive JMX feature (doesn't change existing code). Backport required code changes due to 11u code restructuring and the lack of Thread SMR in 8u, but the changes are confined to additive or additive support code. In production at Amazon for over two months.

**Review by gnu-andrew — APPROVED** [review]
Thanks @simonis for your comments. It made these easier to review. Changes from 11u look fine.

@phohensee where is the 8u CSR? The only one I can find from the bug is the 21u CSR [JDK-8307396](https://bugs.openjdk.org/browse/JDK-8307396) and there seems to be no 8u backport issue.

Given we are a week out from rampdown for 8u412, and this change has not yet been released in 11u & 17u, I would suggest we integrate this after rampdown for 8u422.

**Comment by phohensee** [comment]
@gun-andrew, @jerboaa suggested the same and I'm fine with 8u422.

I've created an 8u backport issue [JDK-8326086](https://bugs.openjdk.org/browse/JDK-8326086). It can use the same CSR as used for 11u and 17u. How do I link a backport JBS issue to a CSR without creating a new one?

**Comment by gnu-andrew** [comment]
> @gun-andrew, @jerboaa suggested the same and I'm fine with 8u422.

Ah ok, I didn't see any comments from him on here.

> 
> I've created an 8u backport issue [JDK-8326086](https://bugs.openjdk.org/browse/JDK-8326086). It can use the same CSR as used for 11u and 17u. How do I link a backport JBS issue to a CSR without creating a new one?

I linked the CSR [JDK-8320375](https://bugs.openjdk.org/browse/JDK-8320375) to the backport bug using the 'csr for' link, which is the same way it's connected to the 17u one. Let's see if it picks that up or not.

**Comment by gnu-andrew** [comment]
> > @gun-andrew, @jerboaa suggested the same and I'm fine with 8u422.
> 
> Ah ok, I didn't see any comments from him on here.
> 
> > I've created an 8u backport issue [JDK-8326086](https://bugs.openjdk.org/browse/JDK-8326086). It can use the same CSR as used for 11u and 17u. How do I link a backport JBS issue to a CSR without creating a new one?
> 
> I linked the CSR [JDK-8320375](https://bugs.openjdk.org/browse/JDK-8320375) to the backport bug using the 'csr for' link, which is the same way it's connected to the 17u one. Let's see if it picks that up or not.

So the CSR needed to be linked to the backport issue and explicitly have 'openjdk8u422' as a version. I think the latter may be worth a Skara bug, as '8-pool' should be sufficient.

**Pull request closed without merging**